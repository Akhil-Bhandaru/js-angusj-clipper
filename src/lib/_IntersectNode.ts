import { CInt } from './types';
import { int } from './_types';
import { TEdge } from './_TEdge';
import { emptyIntPoint, IntPoint } from './IntPoint';

export class IntersectNode {
  Edge1: TEdge; // internal, cant be undefined
  Edge2: TEdge; // internal, cant be undefined
  Pt: IntPoint = emptyIntPoint; // internal
}

export type IntersectNodeComparer = (node1: IntersectNode, node2: IntersectNode) => int;
export const MyIntersectNodeSort: IntersectNodeComparer = (node1: IntersectNode, node2: IntersectNode): int => {
  const i: CInt = node2.Pt.y - node1.Pt.y;
  if (i > 0) return 1;
  else if (i < 0) return -1;
  else return 0;
};
