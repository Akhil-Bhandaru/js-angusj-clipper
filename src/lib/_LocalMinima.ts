import { TEdge } from './_TEdge';
import { CInt } from './types';

export class LocalMinima { // internal
  Y: CInt = 0;
  LeftBound?: TEdge; // can be undefined
  RightBound?: TEdge; // can be undefined
  Next?: LocalMinima; // can be undefined
}
