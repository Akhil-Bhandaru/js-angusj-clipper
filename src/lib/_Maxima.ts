import { CInt } from './types';

export class Maxima { // internal
  X: CInt = 0;
  Next?: Maxima; // can be undefined
  Prev?: Maxima; // can be undefined
}
