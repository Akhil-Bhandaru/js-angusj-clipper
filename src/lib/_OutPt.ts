import { emptyIntPoint, IntPoint } from './IntPoint';

export class OutPt { // internal
  Idx: number = 0;
  Pt: IntPoint = emptyIntPoint;
  Next: OutPt; // cannot be undefined
  Prev: OutPt; // cannot be undefined
}
