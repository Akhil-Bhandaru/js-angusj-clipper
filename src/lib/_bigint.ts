// big int
import {BigInteger} from './BigInteger';

export const Int128Mul = (a: any, b: any): any => {
  return BigInteger.multiply(a, b);
};

export const Int128Equals = (a: any, b: any): boolean => {
  return BigInteger.compareTo(a, b) === 0;
};
