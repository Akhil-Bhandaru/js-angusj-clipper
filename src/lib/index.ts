/*******************************************************************************
 *                                                                              *
 * Author    :  Angus Johnson                                                   *
 * Version   :  6.4.2                                                           *
 * Date      :  27 February 2017                                                *
 * Website   :  http://www.angusj.com                                           *
 * Copyright :  Angus Johnson 2010-2017                                         *
 *                                                                              *
 * License:                                                                     *
 * Use, modification & distribution is subject to Boost Software License Ver 1. *
 * http://www.boost.org/LICENSE_1_0.txt                                         *
 *                                                                              *
 * Attributions:                                                                *
 * The code in this library is an extension of Bala Vatti's clipping algorithm: *
 * "A generic solution to polygon clipping"                                     *
 * Communications of the ACM, Vol 35, Issue 7 (July 1992) pp 56-63.             *
 * http://portal.acm.org/citation.cfm?id=129906                                 *
 *                                                                              *
 * Computer graphics and geometric modeling: implementation and algorithms      *
 * By Max K. Agoston                                                            *
 * Springer; 1 edition (January 4, 2005)                                        *
 * http://books.google.com/books?q=vatti+clipping+agoston                       *
 *                                                                              *
 * See also:                                                                    *
 * "Polygon Offsetting by Computing Winding Numbers"                            *
 * Paper no. DETC2005-85513 pp. 565-575                                         *
 * ASME 2005 International Design Engineering Technical Conferences             *
 * and Computers and Information in Engineering Conference (IDETC/CIE2005)      *
 * September 24-28, 2005 , Long Beach, California, USA                          *
 * http://www.me.berkeley.edu/~mcmains/pubs/DAC05OffsetPolygon.pdf              *
 *                                                                              *
 *******************************************************************************/

/*******************************************************************************
 *                                                                              *
 * Author    :  Javier Gonzalez Garces                                          *
 * Version   :  6.4.2                                                           *
 * Date      :  12 July 2017                                                    *
 *                                                                              *
 * This is a translation of the C# Clipper library to Javascript.               *
 *                                                                              *
 * One change with respect to the original library is that points are assumed   *
 * to be immutable structures. This is done so the algorithm is faster, but if  *
 * you modify a point from the original polygon it might end up modifying the   *
 * result(s).                                                                   *
 *                                                                              *
 * Int128 struct of C# is implemented using js-big-integer                      *
 * Because Javascript lacks support for 64-bit integers, the space              *
 * is a little more restricted than in C# version.                              *
 *                                                                              *
 * C# version has support for coordinate space:                                 *
 * +-4611686018427387903 ( sqrt(2^127 -1)/2 )                                   *
 * while Javascript version has support for space:                              *
 * +-4503599627370495 ( sqrt(2^106 -1)/2 )                                      *
 *                                                                              *
 * js-big-integer proved to be the fastest big integer library for muls:        *
 * http://yaffle.github.io/BigInteger/benchmark/                                *
 *                                                                              *
 * This class can be made simpler when (if ever) 64-bit integer support comes.  *
 *                                                                              *
 *******************************************************************************/

import { CInt, ClipType, PolyType, PolyFillType, JoinType, EndType, Path, Paths,
  PointInPolygonResult, ZFillCallbackImmutable, loRange, hiRange } from './types';
export { CInt, ClipType, PolyType, PolyFillType, JoinType, EndType, Path, Paths,
  PointInPolygonResult, ZFillCallbackImmutable, loRange, hiRange};

import { IntPoint } from './IntPoint';
export { IntPoint };

import { IntRect } from './IntRect';
export { IntRect };

import { PolyNode } from './PolyNode';
export { PolyNode };

import { PolyTree } from './PolyTree';
export { PolyTree };

import { ClipperError } from './ClipperError';
export { ClipperError };

import { area, cleanPolygon, cleanPolygons, openPathsFromPolyTree, closedPathsFromPolyTree,
  minkowskiSumPath, minkowskiSumPaths, minkowskiDiff, polyTreeToPaths, orientation, pointInPolygon,
  reversePath, reversePaths, simplifyPolygon, simplifyPolygons, scalePath, scalePaths } from './functions';
export { area, cleanPolygon, cleanPolygons, openPathsFromPolyTree, closedPathsFromPolyTree,
  minkowskiSumPath, minkowskiSumPaths, minkowskiDiff, polyTreeToPaths, orientation, pointInPolygon,
  reversePath, reversePaths, simplifyPolygon, simplifyPolygons, scalePath, scalePaths };

import { ClipperBase } from './ClipperBase';
export { ClipperBase };

import { Clipper, ClipperInitOptions } from './Clipper';
export { Clipper, ClipperInitOptions };

import { ClipperOffset, ClipperOffsetInitOptions } from './ClipperOffset';
export { ClipperOffset, ClipperOffsetInitOptions };

// TODO: consider moving classes to interfaces to make them faster
