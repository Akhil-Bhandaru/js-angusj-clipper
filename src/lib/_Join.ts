import { emptyIntPoint, IntPoint } from './IntPoint';
import { OutPt } from './_OutPt';

export class Join { // internal
  OutPt1: OutPt; // cannot be undefined
  OutPt2: OutPt; // cannot be undefined
  OffPt: IntPoint = emptyIntPoint;
}
