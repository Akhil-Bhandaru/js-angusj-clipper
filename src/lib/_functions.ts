import { OutPt } from './_OutPt';
import { TEdge } from './_TEdge';
import { double, long, NodeType } from './_types';
import { orientation } from './functions';
import { IntPoint, newIntPointXY, newIntPointXYZ } from './IntPoint';
import { PolyNode } from './PolyNode';
import { Path, Paths } from './types';

export function distanceFromLineSqrd(pt: IntPoint, ln1: IntPoint, ln2: IntPoint): double {
  //The equation of a line in general form (Ax + By + C = 0)
  //given 2 points (x¹,y¹) & (x²,y²) is ...
  //(y¹ - y²)x + (x² - x¹)y + (y² - y¹)x¹ - (x² - x¹)y¹ = 0
  //A = (y¹ - y²); B = (x² - x¹); C = (y² - y¹)x¹ - (x² - x¹)y¹
  //perpendicular distance of point (x³,y³) = (Ax³ + By³ + C)/Sqrt(A² + B²)
  //see http://en.wikipedia.org/wiki/Perpendicular_distance
  const A: double = ln1.y - ln2.y;
  const B: double = ln2.x - ln1.x;
  let C: double = A * ln1.x + B * ln1.y;
  C = A * pt.x + B * pt.y - C;
  return C * C / (A * A + B * B);
}

export function slopesNearCollinear(pt1: IntPoint, pt2: IntPoint, pt3: IntPoint, distSqrd: double): boolean {
  //this function is more accurate when the point that's GEOMETRICALLY
  //between the other 2 points is the one that's tested for distance.
  //nb: with 'spikes', either pt1 or pt3 is geometrically between the other pts
  if (Math.abs(pt1.x - pt2.x) > Math.abs(pt1.y - pt2.y)) {
    if (pt1.x > pt2.x === pt1.x < pt3.x) {
      return distanceFromLineSqrd(pt1, pt2, pt3) < distSqrd;
    }
    else if (pt2.x > pt1.x === pt2.x < pt3.x) {
      return distanceFromLineSqrd(pt2, pt1, pt3) < distSqrd;
    }
    else {
      return distanceFromLineSqrd(pt3, pt1, pt2) < distSqrd;
    }
  }
  else {
    if (pt1.y > pt2.y === pt1.y < pt3.y) {
      return distanceFromLineSqrd(pt1, pt2, pt3) < distSqrd;
    }
    else if (pt2.y > pt1.y === pt2.y < pt3.y) {
      return distanceFromLineSqrd(pt2, pt1, pt3) < distSqrd;
    }
    else {
      return distanceFromLineSqrd(pt3, pt1, pt2) < distSqrd;
    }
  }
}

export function pointsAreClose(pt1: IntPoint, pt2: IntPoint, distSqrd: double): boolean {
  const dx: double = pt1.x - pt2.x;
  const dy: double = pt1.y - pt2.y;
  return dx * dx + dy * dy <= distSqrd;
}

export function excludeOp(op: OutPt): OutPt {
  const result = op.Prev;
  result.Next = op.Next;
  op.Next.Prev = result;
  result.Idx = 0;
  return result;
}

export function addPolyNodeToPaths(polynode: PolyNode, nt: NodeType, paths: Paths): void {
  let match = true;
  switch (nt) {
    case NodeType.ntOpen: return;
    case NodeType.ntClosed:
      match = !polynode.isOpen;
      break;
    default:
      break;
  }

  if (polynode.m_polygon.length > 0 && match) {
    paths.push(polynode.m_polygon);
  }
  for (let ii = 0, max = polynode.childs.length; ii < max; ii++) {
    const pn = polynode.childs[ii];
    addPolyNodeToPaths(pn, nt, paths);
  }
}

export function minkowski(pattern: Path, path: Path, IsSum: boolean, IsClosed: boolean, useXyz: boolean): Paths {
  const delta = IsClosed ? 1 : 0;
  const polyCnt = pattern.length;
  const pathCnt = path.length;
  const result: Paths = [];
  result.length = pathCnt;

  if (IsSum) {
    for (let i = 0; i < pathCnt; i++) {
      const p: Path = [];
      p.length = polyCnt;
      let pLength = 0;
      if (useXyz) {
        for (let ii = 0, max = pattern.length; ii < max; ii++) {
          const ip = pattern[ii];
          p[pLength++] = newIntPointXYZ(path[i].x + ip.x, path[i].y + ip.y, path[i].z! + ip.z!);
        }
      }
      else {
        for (let ii = 0, max = pattern.length; ii < max; ii++) {
          const ip = pattern[ii];
          p[pLength++] = newIntPointXY(path[i].x + ip.x, path[i].y + ip.y);
        }
      }
      //p.length = pLength; // not needed
      result[i] = p;
    }
  }
  else {
    for (let i = 0; i < pathCnt; i++) {
      const p: Path = [];
      p.length = polyCnt;
      let pLength = 0;
      for (let ii = 0, max = pattern.length; ii < max; ii++) {
        const ip = pattern[ii];
        p[pLength++] = newIntPointXYZ(path[i].x - ip.x, path[i].y - ip.y, path[i].z);
      }
      //p.length = pLength; // not needed
      result[i] = p;
    }
  }

  const quads: Paths = [];
  quads.length = (pathCnt - 1 + delta) * (polyCnt); // TODO: originally it is (pathCnt + delta) * (polyCnt + 1) for some reason
  let quadsLength = 0;

  for (let i = 0; i < pathCnt - 1 + delta; i++) {
    for (let j = 0; j < polyCnt; j++) {
      const quad: Path = [
        result[i % pathCnt][j % polyCnt],
        result[(i + 1) % pathCnt][j % polyCnt],
        result[(i + 1) % pathCnt][(j + 1) % polyCnt],
        result[i % pathCnt][(j + 1) % polyCnt],
      ];
      if (!orientation(quad)) {
        quad.reverse();
      }
      quads[quadsLength++] = quad;
    }
  }

  quads.length = quadsLength;
  return quads;
}

export function translatePath(path: Path, delta: IntPoint, useXyz: boolean): Path {
  const outPath: Path = [];
  outPath.length = path.length;
  if (!useXyz) {
    for (let i = 0; i < path.length; i++) {
      outPath[i] = newIntPointXY(path[i].x + delta.x, path[i].y + delta.y);
    }
  }
  else {
    for (let i = 0; i < path.length; i++) {
      outPath[i] = newIntPointXYZ(path[i].x + delta.x, path[i].y + delta.y, path[i].z! + delta.z!);
    }
  }
  return outPath;
}

export const Round = (value: double): long => {
  // TODO: simply use Math.round? although Math.round(-0.5) = 0- and Math.round(0.5) = 1
  return value < 0 ? Math.trunc(value - 0.5) : Math.trunc(value + 0.5);
};

export const TopX = (edge: TEdge, currentY: long) => {
  if (currentY === edge.Top.y) {
    return edge.Top.x;
  }
  return edge.Bot.x + Round(edge.Dx * (currentY - edge.Bot.y));
};

const tolerance: double = 1.0E-20; // internal
export const near_zero = (val: double) => {
  return (val > -tolerance) && (val < tolerance);
};
