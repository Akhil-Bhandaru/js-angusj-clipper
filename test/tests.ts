import * as clipperLib from '../src/lib';

let clipper: clipperLib.ClipperLibWrapper;

beforeAll(async () => {
    clipper = await clipperLib.loadNativeClipperLibInstanceAsync(
        clipperLib.NativeClipperLibRequestedFormat.WasmOnly
    );
});

describe('operations over simple polygons', () => {
    // create some polygons (note that they MUST be integer coordinates)
    const poly1 = [
        { x: 0, y: 10 },
        { x: 30, y: 10 },
        { x: 30, y: 20 },
        { x: 0, y: 20 },
    ];

    const poly2 = [
        { x: 10, y: 0 },
        { x: 20, y: 0 },
        { x: 20, y: 30 },
        { x: 10, y: 30 },
    ];

    for (const clipType of Object.values(clipperLib.ClipType)) {
        for (const polyFillType of [clipperLib.PolyFillType.EvenOdd]) {
            test(`clipType: ${clipType}, subjectFillType: ${polyFillType}`, () => {
                const polyResult = clipper.clipToPaths({
                    clipType: clipType,
            
                    subjectInputs: [
                        { data: poly1, closed: true },
                    ],

                    clipInputs: [
                        { data: poly2 }
                    ],
            
                    subjectFillType: polyFillType
                });
            
                expect(polyResult).toMatchSnapshot();
            });
        }
    }
});
