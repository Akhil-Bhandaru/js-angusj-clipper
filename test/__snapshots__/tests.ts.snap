// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`operations over simple polygons clipType: difference, subjectFillType: evenOdd 1`] = `
Array [
  Array [
    Object {
      "x": 10,
      "y": 20,
    },
    Object {
      "x": 0,
      "y": 20,
    },
    Object {
      "x": 0,
      "y": 10,
    },
    Object {
      "x": 10,
      "y": 10,
    },
  ],
  Array [
    Object {
      "x": 30,
      "y": 20,
    },
    Object {
      "x": 20,
      "y": 20,
    },
    Object {
      "x": 20,
      "y": 10,
    },
    Object {
      "x": 30,
      "y": 10,
    },
  ],
]
`;

exports[`operations over simple polygons clipType: intersection, subjectFillType: evenOdd 1`] = `
Array [
  Array [
    Object {
      "x": 20,
      "y": 20,
    },
    Object {
      "x": 10,
      "y": 20,
    },
    Object {
      "x": 10,
      "y": 10,
    },
    Object {
      "x": 20,
      "y": 10,
    },
  ],
]
`;

exports[`operations over simple polygons clipType: union, subjectFillType: evenOdd 1`] = `
Array [
  Array [
    Object {
      "x": 20,
      "y": 10,
    },
    Object {
      "x": 30,
      "y": 10,
    },
    Object {
      "x": 30,
      "y": 20,
    },
    Object {
      "x": 20,
      "y": 20,
    },
    Object {
      "x": 20,
      "y": 30,
    },
    Object {
      "x": 10,
      "y": 30,
    },
    Object {
      "x": 10,
      "y": 20,
    },
    Object {
      "x": 0,
      "y": 20,
    },
    Object {
      "x": 0,
      "y": 10,
    },
    Object {
      "x": 10,
      "y": 10,
    },
    Object {
      "x": 10,
      "y": 0,
    },
    Object {
      "x": 20,
      "y": 0,
    },
  ],
]
`;

exports[`operations over simple polygons clipType: xor, subjectFillType: evenOdd 1`] = `
Array [
  Array [
    Object {
      "x": 30,
      "y": 20,
    },
    Object {
      "x": 20,
      "y": 20,
    },
    Object {
      "x": 20,
      "y": 30,
    },
    Object {
      "x": 10,
      "y": 30,
    },
    Object {
      "x": 10,
      "y": 20,
    },
    Object {
      "x": 20,
      "y": 20,
    },
    Object {
      "x": 20,
      "y": 10,
    },
    Object {
      "x": 30,
      "y": 10,
    },
  ],
  Array [
    Object {
      "x": 20,
      "y": 10,
    },
    Object {
      "x": 10,
      "y": 10,
    },
    Object {
      "x": 10,
      "y": 20,
    },
    Object {
      "x": 0,
      "y": 20,
    },
    Object {
      "x": 0,
      "y": 10,
    },
    Object {
      "x": 10,
      "y": 10,
    },
    Object {
      "x": 10,
      "y": 0,
    },
    Object {
      "x": 20,
      "y": 0,
    },
  ],
]
`;
